Create Cross-Sells product list

nielslange opened the pull request:
Fixes #6558

#### Accessibility

- [ ] I've tested using only a keyboard (no mouse)
- [ ] I've tested using a screen reader
- [ ] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
- [ ] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](../docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [x] This PR adds/removes an experimental interfaces and I've updated [this doc](../docs/internal-developers/blocks/feature-flags-and-experimental-interfaces.md).
- [ ] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Screenshots

<table>
<tr>
<td>Before:
<br><br>

![6558-before](https://user-images.githubusercontent.com/3323310/176687090-dcbff95e-456f-489f-ab0c-2dddb1603c3d.png)
</td>
<td>After:
<br><br>

![6558-after](https://user-images.githubusercontent.com/3323310/176687105-29613d0a-1baa-4475-a9cb-24624aae3476.png)
</td>
</tr>
</table>

### Testing

#### Test editor behaviour when adding Cart block

1. Create a test page and add the Cart block to it.
2. Verify that the Cart block contains the Cart Cross-Sells inner block.
3. Verify that the Cart Cross-Sells inner block contains the Heading and Cart Cross-Sells products inner blocks.
4. Select the Cart Cross-Sells products inner block and verify that the number of visible Cross-Sells products can be adjusted in the editor.

#### Test editor behaviour after removing Cross-Sells block

1. Create a test page and add the Cart block to it.
2. Verify that the Cart block contains the Cart Cross-Sells inner block.
3. Delete the Cart Cross-Sells inner block.
4. Verify that the Cart Cross-Sells inner block can be reinserted.

#### Test frontend behaviour if cart items does not have Cross-Sells products

1. Visit the frontend.
2. Add a product to the cart that does not Cross-Sells products.
3. Go to the Cart page.
4. Verify that the Cross-Sells product block is not visible.

#### Test frontend behaviour if cart items have Cross-Sells products

1. Visit the frontend.
2. Add a product to the cart that has Cross-Sells products.
3. Go to the Cart page.
4. Verify that the Cross-Sells product block is visible.
5. Add one of the Cross-Sells products, which also has Cross-Sells products, to the cart.
6. Verify that the added Cross-Sells product had been added to the cart and is no longer visible as a Cross-Sells product.
7. Verify that the Cross-Sells products, of the product that had been added to the cart in step 6., are visible in the Cross-Sells block.

#### Test frontend behaviour if cart items have a Cross-Sells product that is out of stock

1. Visit the frontend.
2. Add a product to the cart that has a Cross-Sells product which is out of stock.
3. Go to the Cart page.
4. Verify that the Cross-Sells product block is visible.
5. Verify that the Cross-Sells product, which is out of stock, does not show an `Add to Cart` button, but a `Read More` button, that links to the corresponding product page.

#### Test number of visible Cross-Sells products in the frontend.

1. Ensure that you have one test product which has more than 6 Cross-Sells products.
2. Open the cart page in the editor and set the number of visible Cross-Sells items to 1.
3. Go to the frontend and add the test product from step 1 to the cart.
4. Go to the cart and verify that only 1 Cross-Sells product is visible.
5. Repeat steps 2 until 4 and set the number of visible Cross-Sells items to 2, 3, 4, 5 & 6.

* [ ] Do not include in the Testing Notes <!-- Check this box if this PR can't be tested by users (ie: it doesn't include user-facing changes, or it can't be tested without manually modifying the code). -->

### WooCommerce Visibility

<!-- Check this [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) to see whether the change is visible in WC core or not (part of the feature plugin or experimental)-->

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Performance Impact

n/a

### Changelog

> Create Cross-Sells product list for showing the Cross-Sells products on the Cart block.

tarhi-saad commented on assets/js/blocks/cart/inner-blocks/cart-items-block/edit.tsx:
```diff
@@ -13,11 +14,18 @@ import {
 		[ 'woocommerce/cart-line-items-block', {}, [] ],
+		...( isExperimentalBuild()
+			? [ 'woocommerce/cart-cross-sells-block', {}, [] ]
+			: [] ),
```
Hi @nielslange! You have reported in Slack the following bug: `TypeError: t is not iterable at blocks.min.js?`.  The fix is simple! You just need to wrap the `cross-sells template` inside another array for the `spread operator`, to avoid spreading what's inside this array: `[ 'woocommerce/cart-cross-sells-block', {}, [] ]`:

```js
const defaultTemplate = [
		[ 'woocommerce/cart-line-items-block', {}, [] ],
		...( isExperimentalBuild()
			? [ [ 'woocommerce/cart-cross-sells-block', {}, [] ] ]
			: [] ),
	] as unknown as TemplateArray;
```

The following TS type conversion was necessary as well to fix a TS error: `as unknown as TemplateArray`.

nielslange replied:
Thanks a lot, @tarhi-saad! The fix worked like charm! 🙌

tarhi-saad reviewed (changes requested):
Hi @nielslange! You can find below the fix of the bug you reported on Slack!

wavvves commented:
One thing missing is support for products with options. See attached screenshots for differences between blocks and shortcode on cross-sells:

Block version
![image](https://user-images.githubusercontent.com/17236129/180269131-56cc86ac-6664-4acb-b2c0-e01b31f36992.png)

Shortcode version
![image](https://user-images.githubusercontent.com/17236129/180269233-8bdcf4aa-e014-4b6c-b502-f5b565cc24d8.png)

We can make this an actionable item to work in our pairing session @nielslange 

Also, empty cross-sells still display the heading
![image](https://user-images.githubusercontent.com/17236129/180269628-87d393d1-6447-4281-9619-713e45ede2a8.png)

And we're missing the grey placeholder styles
![image](https://user-images.githubusercontent.com/17236129/180269881-0374844e-bd33-4f7a-930e-2cc79de301d3.png)

nielslange commented on assets/js/blocks/cart/cart-cross-sells-product-list/index.tsx:
```diff
@@ -39,7 +38,7 @@ const CartCrossSellsProductList = ( {
 				);
 		  } );
 
-	return <Noninteractive>{ products }</Noninteractive>;
```
@wavvves I wrapped the products with `<Noninteractive>` to ensure that the whole area can be clicked to select the Cross Sells block in the editor. Removing `<Noninteractive>` leads to the fact that all elements can be clicked and when clicking on the product image or product title, the merchant gets redirected to https://example.org/. 

Wrapping the products with `<Noninteractive>` mimics the behaviour of the Cart Line Items block, which is located above the Cross Sells block.

nielslange committed:
Create Cross-Sells product list

opr commented on assets/js/blocks/cart/cart-cross-sells-product-list/cart-cross-sells-product.tsx:
```diff
@@ -0,0 +1,53 @@
+				parentClassName={ 'wp-block-cart-cross-sells-product' }
+			>
+				<ProductDataContextProvider
+					product={ crossSellsProduct }
```
I see a TS error here: `TS2322: Type 'ProductResponseItem' is not assignable to type 'null | undefined'.`

nielslange replied:
Thanks to @gigitux, the TS problems have been resolved!

opr commented on assets/js/blocks/cart/cart-cross-sells-product-list/cart-cross-sells-product.tsx:
```diff
@@ -0,0 +1,53 @@
+				>
+					<div>
+						<ProductImage />
+						<ProductName />
```
TS Error here: `TS2739: Type '{}' is missing the following properties from type 'Attributes': headingLevel, showProductLink, productId, align` - we may need to fix the type in https://github.com/woocommerce/woocommerce-blocks/blob/c734f3e846b8da7d49a03347d92354efd786251f/assets/js/atomic/blocks/product-elements/title/types.ts#L1

nielslange replied:
See https://github.com/woocommerce/woocommerce-blocks/pull/6645#discussion_r966013853.

opr commented on assets/js/blocks/cart/cart-cross-sells-product-list/cart-cross-sells-product.tsx:
```diff
@@ -0,0 +1,53 @@
+import AddToCartButton from '../../../atomic/blocks/product-elements/add-to-cart/block';
+
+interface CrossSellsProductProps {
+	crossSellsProduct: ProductResponseItem;
```
Maybe it would be better to call this `product`? Does prepending `crossSells` add any value?

nielslange replied:
> Maybe it would be better to call this product? Does prepending crossSells add any value?

I was using `CrossSellsProductProps` instead of `ProductProps` for consistency-reasons, see:

https://github.com/woocommerce/woocommerce-blocks/blob/df63d2ae397b35242c40ea4eec4673c50eaabb15/assets/js/blocks/featured-items/block-controls.tsx#L39

https://github.com/woocommerce/woocommerce-blocks/blob/df63d2ae397b35242c40ea4eec4673c50eaabb15/assets/js/blocks/featured-items/image-editor.tsx#L37

https://github.com/woocommerce/woocommerce-blocks/blob/df63d2ae397b35242c40ea4eec4673c50eaabb15/assets/js/blocks/featured-items/inspector-controls.tsx#L74

https://github.com/woocommerce/woocommerce-blocks/blob/df63d2ae397b35242c40ea4eec4673c50eaabb15/assets/js/blocks/featured-items/with-api-error.tsx#L22

https://github.com/woocommerce/woocommerce-blocks/blob/df63d2ae397b35242c40ea4eec4673c50eaabb15/assets/js/blocks/featured-items/with-featured-item.tsx#L91

https://github.com/woocommerce/woocommerce-blocks/blob/18c8e4597bf0225fd67ba63f854a2448e2498f54/assets/js/blocks/featured-items/with-update-button-attributes.tsx#L29

That said, if you believe that `ProductProps` is a better naming, just let me know.

opr replied:
In all the examples you linked, the product is still named `product` - this one is the odd-one-out because it's called `crossSellsProduct`

opr replied:
By the way I was referring to `crossSellsProduct` on line 21 here.

nielslange replied:
Thanks for adding more context. I thought you were referring to line 20. I just implemented your suggestion.

opr commented on assets/js/blocks/cart/cart-cross-sells-product-list/index.tsx:
```diff
@@ -0,0 +1,44 @@
+import CartCrossSellsProduct from './cart-cross-sells-product';
+
+const placeholderRows = [ ...Array( 3 ) ].map( ( _x, i ) => (
+	<CartCrossSellsProduct crossSellsProduct={ {} } key={ i } />
```
This doesn't seem to do anything, I replaced it with placeholder images and they are never rendered. Also, I'm curious if we considered creating a separate component called `CrossSellsProductPlaceholder` or something instead of a `CartCrossSellsProduct` with no data?

I see `CrossSellsProduct` has an `isLoading` prop, but this seems to be unused?

nielslange replied:
I removed both `isLoading` and `placeholderRows`, given that the Cross-Sells product section will only appear if the products in the cart have Cross-Sells products.

opr commented on assets/js/blocks/cart/cart-cross-sells-product-list/index.tsx:
```diff
@@ -0,0 +1,44 @@
+}
+
+const CartCrossSellsProductList = ( {
+	crossSellsProducts,
```
Would `products` be a better name? the `crossSells` doesn't seem to add much value.

nielslange replied:
See the comment I added to https://github.com/woocommerce/woocommerce-blocks/pull/6645/files/aed95be381fedeb379eabc460a78e591b8106a61#r956142844.

opr replied:
I don't think that comment applies here, I was referring to the prop `crossSellsProducts` - `products` seems like it would be fine?

nielslange replied:
I see now what you meant and just changed it! Thanks for pointing this out, @opr.

opr commented on assets/js/blocks/cart/cart-cross-sells-product-list/index.tsx:
```diff
@@ -0,0 +1,44 @@
+
+const CartCrossSellsProductList = ( {
+	crossSellsProducts,
+	isLoading = false,
```
I never saw this be `true`, could it be because our block is only rendered _after_ the initial cart loading work is complete?

nielslange replied:
That's possible, and I removed this part as a result of https://github.com/woocommerce/woocommerce-blocks/pull/6645/files/aed95be381fedeb379eabc460a78e591b8106a61#r956144991.

opr commented on assets/js/blocks/cart/inner-blocks/cart-cross-sells-block/frontend.tsx:
```diff
@@ -0,0 +1,24 @@
+	children,
+	className = '',
+}: Props ): JSX.Element | null => {
+	const { crossSellsProducts, cartIsLoading } = useStoreCart();
```
Let's investigate why `cartIsLoading` is always false for our block, it could be because it's an inner block and it's only rendered when this is finished, if so we could remove this.

nielslange replied:
That would be great, if you could assist me troubleshooting this one.

opr commented on assets/js/blocks/cart/inner-blocks/cart-cross-sells-products/block.tsx:
```diff
@@ -0,0 +1,34 @@
+	const { crossSellsProducts, cartIsLoading } = useStoreCart();
+
+	if ( typeof columns === 'undefined' ) {
+		columns = metadata.attributes.columns.default;
```
Are we guaranteed to always have this, would `metadata?.attributes?.columns?.default` be safer or is it unnecessary?

nielslange replied:
> Are we guaranteed to always have this [...]?

Based on the following section in block.json, I was under the impression that `metadata.attributes.columns.default` always exists:

```json
"attributes": {
  "columns": {
    "type": "number",
    "default": 3
  }
},
```

Am I missing something in this case?

opr replied:
What I meant was: is there any case where `metadata` or any of its children could be undefined, or a non-object. Looking at the rest of the codebase for `metadata.attributes`, it seems we're not worried about that. Marking as resolved!

opr commented on assets/js/blocks/cart/cart-cross-sells-product-list/index.tsx:
```diff
@@ -0,0 +1,44 @@
+interface CrossSellsProductListProps {
+	crossSellsProducts: ProductResponseItem[];
+	isLoading: boolean;
+	className?: string;
```
Because of https://github.com/woocommerce/woocommerce-blocks/blob/trunk/tsconfig.base.json#L10 we need to allow undefined here.

```suggestion
	className?: string | undefined;
```

opr commented on assets/js/blocks/cart/inner-blocks/cart-cross-sells-products/block.tsx:
```diff
@@ -0,0 +1,34 @@
+import metadata from './block.json';
+
+interface BlockProps {
+	className?: string;
```
Because of https://github.com/woocommerce/woocommerce-blocks/blob/trunk/tsconfig.base.json#L10 we need to allow undefined here.

```suggestion
	className?: string | undefined;
```

opr commented on assets/js/blocks/cart/style.scss:
```diff
@@ -21,7 +21,7 @@ table.wc-block-cart-items td {
 	// Remove borders on default themes.
 	border: 0;
-	margin: 0;
+	margin: 0 0 2em;
```
This margin will be applied regardless of the cross-sells block's presence, right? Are we OK with this?

nielslange replied:
I'd say it's safe to keep this even when the Cross-Sells block is not visible, as it ensures that there's spacing below the cart items block. For example, if a merchant decides adding some content below the cart items block.

<table>
<tr>
<td>With the spacing above
<br><br>
<img width="854" alt="Screenshot 2022-08-27 at 12 07 46" src="https://user-images.githubusercontent.com/3323310/187025569-47af3d90-7fe7-4b0d-a658-bc182676dd76.png">
</td>
<td>Without the spacing above
<br><br>
<img width="849" alt="Screenshot 2022-08-27 at 12 08 23" src="https://user-images.githubusercontent.com/3323310/187025599-9e579069-efd2-41f1-b8a0-f98c222dfee8.png">
</td>
</tr>
</table>



While I introduced that change for the Cross-Sells block, other elements equally benefit from this change.

opr commented on assets/js/types/type-defs/cart.ts:
```diff
@@ -190,6 +193,7 @@ export interface Cart {
 	items: Array< CartItem >;
 	itemsCount: number;
 	itemsWeight: number;
+	crossSells: Array< ProductResponseItem >;
```
Is this true, we're storing `crossSells` in the same format as the API gives back, but `items` [get parsed into camelCase](https://github.com/woocommerce/woocommerce-blocks/blob/df63d2ae397b35242c40ea4eec4673c50eaabb15/assets/js/data/cart/actions.ts#L37) - maybe we could do the same for `crossSells`?

nielslange replied:
I'm not sure what exactly you mean, @opr. Would you mind explaining it to me?

opr replied:
Oops, never mind, I was misunderstanding how the cart got processed from the sever side, I thought the same happened to items in it, but it doesn't. This is fine.

opr reviewed (changes requested):
Hey @niels, good work. This looks really good!

All of your testing steps passed, but I did a few more and found some issues:

1. When cross-sells block has one product in, adding that product to the cart results in `Warning: Can't perform a React state update on an unmounted component`, seems related to `AddToCartFormStateContextProvider` so may be unrelated, and could possibly be addressed in a separate PR.
2. Out of stock items appear in the list and the user can try to add them to the cart. It does not get added and no errors are presented to the user, however in console you can see the request failing with a 400 error.

I've also left some inline comments which would be great to address. Let me know if you need help with clarify anything I wrote! 🍻

nielslange committed:
Show “Read more” button for out-of-stock cross-sells products

nielslange committed:
Merge branch 'trunk' into add/6558-create-cross-sells-product-list

nielslange committed:
Update assets/js/blocks/cart/inner-blocks/cart-cross-sells-products/block.tsx

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

nielslange committed:
Update assets/js/blocks/cart/cart-cross-sells-product-list/index.tsx

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

wavvves commented:
> 2\. Out of stock items appear in the list and the user can try to add them to the cart. It does not get added and no errors are presented to the user, however in console you can see the request failing with a 400 error.

Hi @opr this may be related to the StoreApi endpoint and probably needs to get looked at separately. Btw, did you compare it with the shortcode version?

opr commented:
>  Btw, did you compare it with the shortcode version?

Good point @wavvves, no I didn't! @nielslange could you check it out when you come back to this?

nielslange committed:
Merge branch 'trunk' into add/6558-create-cross-sells-product-list

nielslange committed:
Remove obsolete isLoading and placeholderRows

nielslange committed:
Merge branch 'add/6558-create-cross-sells-product-list' of github.com:woocommerce/woocommerce-blocks into add/6558-create-cross-sells-product-list

nielslange reviewed (commented):
Thanks for your review, @opr. I fixed what I could, add some notes regarding your questions and might need some help with the two TS errors as well as looking into the `cartIsLoading` issue.

nielslange commented:
> > 2. Out of stock items appear in the list and the user can try to add them to the cart. It does not get added and no errors are presented to the user, however in console you can see the request failing with a 400 error.
> 
> Hi @opr this may be related to the StoreApi endpoint and probably needs to get looked at separately. Btw, did you compare it with the shortcode version?

@wavvves and @opr I compared this issue with the shortcode-based solution and noticed that only the button in the Cross-Sells block didn't work as expected. I went ahead and fixed that. For products, that are out-of-stock, the button no longer says "Add to cart" but "Read more". Now, both the shortcode-based solution and the Cross-Sells block behave identical for out-of-stock products.

nielslange committed:
Merge branch 'trunk' into add/6558-create-cross-sells-product-list

nielslange committed:
Fix TS errors

nielslange committed:
Merge branch 'trunk' into add/6558-create-cross-sells-product-list

nielslange committed:
Rename crossSellsProduct to product

nielslange committed:
Merge branch 'add/6558-create-cross-sells-product-list' of github.com:woocommerce/woocommerce-blocks into add/6558-create-cross-sells-product-list

nielslange committed:
Merge branch 'trunk' into add/6558-create-cross-sells-product-list

nielslange committed:
Merge branch 'trunk' into add/6558-create-cross-sells-product-list

nielslange committed:
Merge branch 'trunk' into add/6558-create-cross-sells-product-list

tarhi-saad reviewed (commented):
Thank you, @nielslange, for working on this great PR! 🙌 

This is the first time I'm testing the PR, so I'm not sure if I did something wrong, but, I'm getting an error on the editor side of the Cart Block:

<img width="551" alt="image" src="https://user-images.githubusercontent.com/14235870/190026774-52d49a82-1c00-4d4a-8eb0-69cefbf18429.png">

And on the front-end I'm getting the following error:

<img width="666" alt="image" src="https://user-images.githubusercontent.com/14235870/190027820-6b07c5e5-bef4-45a8-8bf2-54a9538d251a.png">

nielslange committed:
Merge branch 'trunk' into add/6558-create-cross-sells-product-list

nielslange committed:
Fix critical error

nielslange commented:
> This is the first time I'm testing the PR, so I'm not sure if I did something wrong, but, I'm getting an error on the editor side of the Cart Block:

@tarhi-saad I double-checked the PR and noticed that the error occurred due to renaming variables in https://github.com/woocommerce/woocommerce-blocks/pull/6645/commits/8f1c545b0f860e2a3df27382ee71afc463543612. I just resolved that problem. The Cross-Sells block should now appear, as expected, both in the editor and the frontend.

opr reviewed (commented):
Thanks @nielslange I tested and it works great - I'll mention something I wondered about. When it comes to removing the block, if you remove the actual block and the heading above it  the block still remains, is there anything we can do about this?

Also, the block reappeared once when I removed the product list and heading. Could you look into why this is happening? (You can see this on the video below)

https://www.loom.com/share/c09a3e36a5504cb9bbd26651ca27c5dc

nielslange committed:
Merge branch 'trunk' into add/6558-create-cross-sells-product-list

nielslange committed:
Merge branch 'trunk' into add/6558-create-cross-sells-product-list

nielslange commented:
> Thanks @nielslange I tested and it works great - I'll mention something I wondered about. When it comes to removing the block, if you remove the actual block and the heading above it the block still remains, is there anything we can do about this?
> 
> Also, the block reappeared once when I removed the product list and heading. Could you look into why this is happening? (You can see this on the video below)
> 
> https://www.loom.com/share/c09a3e36a5504cb9bbd26651ca27c5dc

Thanks for your review and pointing this out, @opr. 

> When it comes to removing the block, if you remove the actual block and the heading above it the block still remains, is there anything we can do about this?

I looked into options, but cannot find anything for now. That said, I noticed that Gutenberg's Group block and our Product Query and Products on Sale blocks behave similar. When deleting inner blocks, both siblings and the parent block remain unaffected. In addition, the parent block does not get deleted if it has no more inner blocks. I agree that this is not ideal, but I'd rather look into this issue in a follow-up issue. Given that this problem also appears when using Gutenberg's Group block, this issue might be addressed in upstream.

> Also, the block reappeared once when I removed the product list and heading. Could you look into why this is happening? 
I'm unable to reproduce this issue on my end, both when using an existing site and a brand-new site.

tarhi-saad reviewed (changes requested):
Thank you, @nielslange, for working on this PR! I've gone through all the tests mentioned above, and everything is working as expected! 🙌

> I looked into options, but cannot find anything for now. That said, I noticed that Gutenberg's Group block and our Product Query and Products on Sale blocks behave similar. When deleting inner blocks, both siblings and the parent block remain unaffected. In addition, the parent block does not get deleted if it has no more inner blocks. I agree that this is not ideal, but I'd rather look into this issue in a follow-up issue. Given that this problem also appears when using Gutenberg's Group block, this issue might be addressed in upstream.

I believe what makes the Cross-Sells different than some of Gutenberg's blocks, is that we prevent the user from adding the `Cross-Sells product list` block if the parent isn't deleted. This can be frustrating to the user wondering why he isn't allowed to add the block again after removing it using the Editor UI (see image below):

<img width="578" alt="image" src="https://user-images.githubusercontent.com/14235870/191148006-ad7c75e3-02fc-4355-8f52-642c99a7753b.png">

The image above shows that there is no `Cross-Sells product lits` block, nonetheless, we can't add one! The only way to fix the issue will be for the user to open the `List View` and notice the parent block is still there.

If there isn't an easy solution to remove the parent when removing its inner block, can we look into preventing removing the inner block itself, and only allowing removing the block by deleting the parent? I believe there is a Gutenberg block that works this way.

> Also, the block reappeared once when I removed the product list and heading. Could you look into why this is happening? (You can see this on the video below)

I was able to replicate this bug on my end as well. It only happens once after a page refresh, so to reproduce the same bug you'll need to refresh the page and try again:

1. Go to Cart with the Cross-sell block
2. Remove the `Cart Cross-Sells products` inner block first (using the UI or through the List View).
3. Then remove the `Heading` inner block. Both inner blocks will show up again!
4. If the inner blocks were removed as expected, refresh the page and try again!

Lastly, I noticed that the `Cart Cross-Sells block` doesn't show for the existing Cart Blocks. We should remove our existing Cart and add it again to work. Is this an expected behavior? To reproduce:

1. Checkout to the `trunk` branch
2. Execute `npm run build`
3. Create a new page and add a Cart Block
4. Checkout to this branch (i.e., `add/6558-create-cross-sells-product-list`)
5. Execute `npm run build`
6. refresh the newly create Cart page. Check there isn't a `Cart Cross-Sells block`

nielslange committed:
Merge branch 'trunk' into add/6558-create-cross-sells-product-list

nielslange committed:
Lock “Cart Cross-Sells products” inner block

nielslange committed:
Merge branch 'add/6558-create-cross-sells-product-list' of github.com:woocommerce/woocommerce-blocks into add/6558-create-cross-sells-product-list

nielslange commented:
Thanks for your review, @tarhi-saad!

> If there isn't an easy solution to remove the parent when removing its inner block, can we look into preventing removing the inner block itself, and only allowing removing the block by deleting the parent? I believe there is a Gutenberg block that works this way.

I also had that solution in mind, but initially decided against it, as I found the usability too confusing. That said, I just locked the Cross-Sells products inner block. That way, a merchant can either delete the headline or the whole Cross-Sells block, but not the Cross-Sells products inner block.

> I was able to replicate this bug on my end as well. It only happens once after a page refresh, so to reproduce the same bug you'll need to refresh the page and try again:
> 
> 1. Go to Cart with the Cross-sell block
> 2. Remove the Cart Cross-Sells products inner block first (using the UI or through the List View).
> 3. Then remove the Heading inner block. Both inner blocks will show up again!
> 4. If the inner blocks were removed as expected, refresh the page and try again!

Apparently, by locking the Cross-Sells products inner block, this problem seems to have been solved as well. At least, I can no longer reproduce it, while I was able to reproduce it, using your steps, before the inner block was locked.

> Lastly, I noticed that the Cart Cross-Sells block doesn't show for the existing Cart Blocks. We should remove our existing Cart and add it again to work. Is this an expected behavior? To reproduce:
> 
> 1. Checkout to the trunk branch
> 2. Execute npm run build
> 3. Create a new page and add a Cart Block
> 4. Checkout to this branch (i.e., add/6558-create-cross-sells-product-list)
> 5. Execute npm run build
> 6. refresh the newly create Cart page. Check there isn't a Cart Cross-Sells block

That's intentional and was one of [the project deliverables](https://woocheckout.wordpress.com/2022/05/23/cross-sell-block-prd/#project-deliverable).

Would you mind reviewing this PR again, @tarhi-saad ?

tarhi-saad commented on assets/js/blocks/cart/inner-blocks/cart-cross-sells-products/block.json:
```diff
@@ -0,0 +1,30 @@
+		"html": false,
+		"multiple": false,
+		"reusable": false,
+		"inserter": false
```
Hey @nielslange! Can we remove the option to unlock this inner block (e.g; the same way it's done for our other inner blocks)? The reason for applying this restriction is after unlocking the `Cart Cross-Sells products` inner block, the previous bugs can be reproduced again.

<table>
<tr>
<td>Before:
<br><br>

<img width="571" alt="image" src="https://user-images.githubusercontent.com/14235870/191275179-2fc46904-375a-4937-9496-c99bfc377958.png">
</td>
<td>After:
<br><br>

<img width="571" alt="image" src="https://user-images.githubusercontent.com/14235870/191273941-5e8e0bd2-ba1a-4b20-a219-cd1d36c31719.png">
</td>
</tr>
</table>

Below you will find a code suggestion that you can commit if you agree with this change! 👍 

```suggestion
		"inserter": false,
		"lock": false
```

nielslange replied:
Great suggestion! I just committed your suggestion, @tarhi-saad.

tarhi-saad commented on assets/js/blocks/cart/inner-blocks/cart-cross-sells-products/block.json:
```diff
@@ -0,0 +1,30 @@
+			"type": "object",
+			"default": {
+				"remove": true,
+				"move": true
```
If you prefer we can let the user move the `Cart Cross-Sells products` inner block even if it's locked using the following change:

```suggestion
				"move": false
```

nielslange replied:
Thanks for suggesting this. To ensure the same layout of the Cross-Sells block as seen on https://woocommerce.com/document/related-products-up-sells-and-cross-sells/#section-2, I'd keep this lock active for now.

tarhi-saad reviewed (approved):
Thank you for the clarification, @nielslange! I've left some comments below, but nothing crucial! It's just things good to have or debatable! So, feel free to merge the PR after this! 🙌

nielslange committed:
Merge branch 'trunk' into add/6558-create-cross-sells-product-list

nielslange committed:
Update assets/js/blocks/cart/inner-blocks/cart-cross-sells-products/block.json

Co-authored-by: Saad Tarhi <saad.trh@gmail.com>

nielslange committed:
Prevent moving of the Cross-Sells block
